src/idpi/grib_decoder.py

```diff
@@ -82,6 +52,171 @@ def _extract_pv(pv):
     }
 
 
+@dc.dataclass
+class _FieldBuffer:
+    dims: tuple[str, ...] | None = None
+    hcoords: dict[str, xr.DataArray] = dc.field(default_factory=dict)
+    metadata: dict[str, typing.Any] = dc.field(default_factory=dict)
+    time_meta: dict[int, dict] = dc.field(default_factory=dict)
+    values: dict[tuple[int, ...], np.ndarray] = dc.field(default_factory=dict)
+
+    def load(self, field):
+        dim_keys = (
+            ("perturbationNumber", "step", "level")
+            if _is_ensemble(field)
+            else ("step", "level")
+        )
+        key = field.metadata(*dim_keys)
+        logger.debug("Received field for key: %s", key)
+        self.values[key] = field.to_numpy(dtype=np.float32)
+
+        step = key[-2]  # assume all members share the same time steps
+        if step not in self.time_meta:
+            self.time_meta[step] = field.metadata(namespace="time")
+
+        if not self.dims:
+            self.dims = tuple(DIM_MAP[d] for d in dim_keys) + ("y", "x")
+
+        if not self.metadata:
+            self.metadata = {
+                "message": field.message(),
+                **metadata.extract(field.metadata()),
+            }
+
+        if not self.hcoords:
+            self.hcoords = {
+                dim: xr.DataArray(dims=("y", "x"), data=values)
+                for dim, values in field.to_latlon().items()
+            }
+
+    def _gather_coords(self):
+        if self.dims is None:
+            raise RuntimeError("No dims.")
+
+        coord_values = zip(*self.values)
+        unique = (sorted(set(values)) for values in coord_values)
+        coords = {dim: c for dim, c in zip(self.dims[:-2], unique)}
+
+        if missing := [
+            combination
+            for combination in product(*coords.values())
+            if combination not in self.values
+        ]:
+            msg = f"Missing combinations: {missing}"
+            logger.exception(msg)
+            raise RuntimeError(msg)
+
+        ny, nx = next(iter(self.values.values())).shape
+        shape = tuple(len(v) for v in coords.values()) + (ny, nx)
+        return coords, shape
+
+    def _gather_tcoords(self):
+        time = None
+        valid_time = []
+        for step in sorted(self.time_meta):
+            tm = self.time_meta[step]
+            valid_time.append(_parse_datetime(tm["validityDate"], tm["validityTime"]))
+            if time is None:
+                time = _parse_datetime(tm["dataDate"], tm["dataTime"])
+
+        return {"valid_time": ("time", valid_time), "ref_time": time}
+
+    def to_xarray(self) -> xr.DataArray:
+        if not self.values:
+            raise RuntimeError("No values.")
+
+        coords, shape = self._gather_coords()
+        tcoords = self._gather_tcoords()
+
+        array = xr.DataArray(
+            data=np.array(
+                [self.values.pop(key) for key in sorted(self.values)]
```

Comment:
To avoid modifying `self.values` by popping its values, could we rather use the following here below?
```
np.array(
    [self.values[key] for key in sorted(self.values)]
)
```

Reply:
That was the initial implementation but the pop was introduced to avoid holding 2 copies of the data which effectively doubles the peak memory footprint.